libraries/chain/include/graphene/chain/custom_authority_object.hpp

```diff
@@ -0,0 +1,119 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#pragma once
+#include <graphene/protocol/authority.hpp>
+#include <graphene/protocol/custom_authority.hpp>
+#include <graphene/protocol/restriction_predicate.hpp>
+#include <graphene/db/generic_index.hpp>
+#include <graphene/chain/types.hpp>
+#include <boost/multi_index/composite_key.hpp>
+
+namespace graphene { namespace chain {
+
+   /**
+    * @brief Tracks account custom authorities
+    * @ingroup object
+    *
+    */
+   class custom_authority_object : public abstract_object<custom_authority_object> {
+      /// Unreflected field to store a cache of the predicate function
+      /// Note that this cache can be modified when the object is const!
+      optional<restriction_predicate_function> predicate_cache;
+
+   public:
+      static const uint8_t space_id = protocol_ids;
+      static const uint8_t type_id = custom_authority_object_type;
+
+      account_id_type account;
+      bool enabled;
+      time_point_sec valid_from;
+      time_point_sec valid_to;
+      unsigned_int operation_type;
+      authority auth;
+      flat_map<uint16_t, restriction> restrictions;
+      uint16_t restriction_counter = 0;
+
+      /// Check whether the custom authority is valid
+      bool is_valid(time_point_sec now) const { return enabled && now >= valid_from && now < valid_to; }
+
+      /// Get the restrictions as a vector rather than a map
+      vector<restriction> get_restrictions() const {
+         vector<restriction> rs;
+         std::transform(restrictions.begin(), restrictions.end(),
+                        std::back_inserter(rs), [](auto i) { return i.second; });
+         return rs;
+      }
+      /// Get predicate, from cache if possible, and update cache if not (modifies const object!)
+      restriction_predicate_function get_predicate() const {
+         if (predicate_cache.valid())
+            return *predicate_cache;
+
+         const_cast<custom_authority_object*>(this)->update_predicate_cache();
```

Comment:
So that only `update_predicate_cache()` can modify the field, but that method is not `const`. Do you think it would be better to make `update_predicate_cache()` a `const` method, and make the field mutable?

Reply:
IMO the cache is an implementation detail of the class. Unfortunately it cannot be hidden completely, because thanks to our reflection mechanism the `restrictions` field is public and can be written by anyone instead of going through a setter that would invalidate the cache below the hood.
It may be a matter of taste if `const_cast` or `mutable` is better here. Perhaps the distinction becomes a little more clear with a `clear_cache` method - this only needs to be called when `restrictions` has been modified, which requires a non-const instance anyway. `update_cache`, however, stays an implementation detail - the caller of `get_predicate` need not worry if the cache gets updated or not, and will typically call it on a const instance. IMO that's exactly what `mutable` is meant to express - to an external observer the cache is not part of the actual const object state.

Reply:
So do you recommend making the cache `mutable` and `update_cache` `const`?

Reply:
yes

Reply:
OK, done